pyquil/api/_base_connection.py

```diff
@@ -0,0 +1,183 @@
+##############################################################################
+# Copyright 2016-2017 Rigetti Computing
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+##############################################################################
+
+from __future__ import print_function
+
+import requests
+import sys
+
+import time
+from requests.adapters import HTTPAdapter
+from six import integer_types
+from urllib3 import Retry
+
+from .job import Job
+from ._config import PyquilConfig
+
+TYPE_EXPECTATION = "expectation"
+TYPE_MULTISHOT = "multishot"
+TYPE_MULTISHOT_MEASURE = "multishot-measure"
+TYPE_WAVEFUNCTION = "wavefunction"
+
+
+class BaseConnection(object):
+    def __init__(self, endpoint, api_key=None, user_id=None):
+        self._session = requests.Session()
+        retry_adapter = HTTPAdapter(max_retries=Retry(total=3,
+                                                      method_whitelist=['POST'],
+                                                      status_forcelist=[502, 503, 504, 521, 523],
+                                                      backoff_factor=0.2,
+                                                      raise_on_status=False))
+
+        # We need this to get binary payload for the wavefunction call.
+        self._session.headers.update({"Accept": "application/octet-stream"})
+
+        self._session.mount("http://", retry_adapter)
+        self._session.mount("https://", retry_adapter)
+
+        config = PyquilConfig()
+        self.endpoint = endpoint
```

Comment:
If I have `api.rigetti.com` in my config file, won't this overwrite the endpoint that is set by default for the QPUConnection?  e.g. if my pyquil_config is setting api.rigetti.com and I call `qpu = QPUConnection()` this will have the wrong endpoint?

Reply:
This could be fixed if we tell all users to have the job url.  Is that the way to go?  Likely most users have api.rigetti.com right now in their config files as that has been the default

Reply:
Good question: in an earlier commit I removed the functionality of reading the endpoint from the config file. Everyone had the same thing anywhere, so it made more sense to just make it a constant. Now the average user will never have to think about endpoints at all.

Reply:
Yes, this is the way to go 👍